src/services/review/ReviewRequestService.ts

```diff
@@ -71,70 +100,163 @@ export default class ReviewRequestService {
+      const extractedPathResult = extractPathInfo(filename)
+      if (extractedPathResult.isErr()) {
+        return errAsync({
+          name: "",
+          path: [],
+          type: "file",
+        })
+      }
 
-    const mappings = await this.computeShaMappings(commits)
+      const pathInfo = extractedPathResult.value
 
-    return Promise.all(
-      files.map(async ({ filename, contents_url }) => {
-        const fullPath = filename.split("/")
-        const items = contents_url.split("?ref=")
-        // NOTE: We have to compute sha this way rather than
-        // taking the file sha.
-        // This is because the sha present on the file is
-        // a checksum of the files contents.
-        // And the actual commit sha is given by the ref param
-        const sha = items[items.length - 1]
-        const pageNameRes = this.pageService.parsePageName(
-          filename,
-          sessionData
+      return this.extractConfigInfo(pathInfo)
+        .orElse(() => this.extractMediaInfo(pathInfo))
+        .asyncMap<EditedItemDto>(async (item) => item)
+        .orElse(() =>
+          this.extractPageInfo(pathInfo, sessionData, stagingLink, siteName)
         )
-        const stagingUrl = await pageNameRes
-          .andThen((pageName) =>
-            this.pageService.retrieveStagingPermalink(
-              sessionData,
-              stagingLink,
-              pageName
-            )
-          )
-          // NOTE: We ignore the errors and use a placeholder
-          .unwrapOr("")
-        const fileUrl = await pageNameRes
-          .andThen((pageName) =>
-            this.pageService.retrieveCmsPermalink(pageName, siteName)
-          )
-          .unwrapOr("")
+        .map<WithEditMeta<EditedItemDto>>((item) => {
+          const items = contents.split("?ref=")
+          // NOTE: We have to compute sha this way rather than
+          // taking the file sha.
+          // This is because the sha present on the file is
+          // a checksum of the files contents.
+          // And the actual commit sha is given by the ref param
+          const hash = items[items.length - 1]
+          return {
+            ...item,
+            lastEditedBy: mappings[hash]?.author || "Unknown user",
+            lastEditedTime: mappings[hash]?.unixTime || 0,
+          }
+        })
 
-        return {
-          type: this.computeFileType(filename),
-          // NOTE: The string is guaranteed to be non-empty
-          // and hence this should exist.
-          name: fullPath.pop()!,
-          // NOTE: pop alters in place
-          path: fullPath,
-          stagingUrl,
-          fileUrl,
-          lastEditedBy: mappings[sha]?.author || "Unknown user",
-          lastEditedTime: mappings[sha]?.unixTime || 0,
-        }
+        .orElse<BaseEditedItemDto>(() => {
+          const { path, name } = pathInfo
+          return errAsync({
+            name,
+            path: path.unwrapOr([]),
```

Comment:
sorry can i check where does this unWrapOr come from? what does it do?

Reply:
`neverthrow` - if it's ok, returns the value in ok, otherwise uses a default

Reply:
Ooh okay, I think this syntax is going take some time to get used to. Do you think we need neverthrow though? I thought of discussing with you how best to standardise returns and responses across the codebase first

Reply:
shure, can!